Mathlib/Topology/UnitInterval.lean

```diff
@@ -179,6 +193,87 @@ theorem two_mul_sub_one_mem_iff {t : ℝ} : 2 * t - 1 ∈ I ↔ t ∈ Set.Icc (1
 
 end unitInterval
 
+section partition
+
+namespace Set.Icc
+
+variable {α} [LinearOrderedAddCommGroup α] {a b c d : α} (h : a ≤ b) {δ : α}
+
+-- TODO: Set.projIci, Set.projIic
+/-- `Set.projIcc` is a contraction. -/
+lemma _root_.Set.abs_projIcc_sub_projIcc : (|projIcc a b h c - projIcc a b h d| : α) ≤ |c - d| := by
+  wlog hdc : d ≤ c generalizing c d
+  · rw [abs_sub_comm, abs_sub_comm c]; exact this (le_of_not_le hdc)
+  rw [abs_eq_self.2 (sub_nonneg.2 hdc), abs_eq_self.2 (sub_nonneg.2 <| monotone_projIcc h hdc)]
+  rw [← sub_nonneg] at hdc
+  refine (max_sub_max_le_max _ _ _ _).trans (max_le (by rwa [sub_self]) ?_)
+  refine ((le_abs_self _).trans <| abs_min_sub_min_le_max _ _ _ _).trans (max_le ?_ ?_)
+  · rwa [sub_self, abs_zero]
+  · exact (abs_eq_self.mpr hdc).le
+
+/-- Equally spaced points in a closed interval. -/
+def addNsmul (δ : α) (n : ℕ) : Icc a b := projIcc a b h (a + n • δ)
+
+lemma addNsmul_zero : addNsmul h δ 0 = a := by
+  rw [addNsmul, zero_smul, add_zero, projIcc_left]
+
+lemma addNsmul_eq_right [Archimedean α] (hδ : 0 < δ) :
+    ∃ m, ∀ n ≥ m, addNsmul h δ n = b := by
+  obtain ⟨m, hm⟩ := Archimedean.arch (b - a) hδ
+  refine ⟨m, fun n hn ↦ ?_⟩
+  rw [addNsmul, coe_projIcc, add_comm, min_eq_left_iff.mpr, max_eq_right h]
+  exact sub_le_iff_le_add.mp (hm.trans <| nsmul_le_nsmul hδ.le hn)
+
+lemma monotone_addNsmul (hδ : 0 ≤ δ) : Monotone (addNsmul h δ) :=
+  fun _ _ hnm ↦ monotone_projIcc h <| (add_le_add_iff_left _).mpr (nsmul_le_nsmul hδ hnm)
+
+lemma abs_sub_addNsmul_le (hδ : 0 ≤ δ) {t : Icc a b} (n : ℕ)
+    (ht : t ∈ Icc (addNsmul h δ n) (addNsmul h δ (n+1))) :
+    (|t - addNsmul h δ n| : α) ≤ δ :=
+  (abs_eq_self.2 <| sub_nonneg.2 ht.1).trans_le <| (sub_le_sub_right (by exact ht.2) _).trans <|
+    (le_abs_self _).trans <| (abs_projIcc_sub_projIcc h).trans <| by
+      rw [add_sub_add_comm, sub_self, zero_add, succ_nsmul, add_sub_cancel]
+      exact (abs_eq_self.mpr hδ).le
+
+end Set.Icc
+
+open scoped unitInterval
+
+/-- Any open cover of a closed interval in ℝ can be refined to
+  a finite partition into subintervals. -/
+lemma lebesgue_number_lemma_Icc {ι} {a b : ℝ} (h : a ≤ b) {c : ι → Set (Icc a b)}
+    (hc₁ : ∀ i, IsOpen (c i)) (hc₂ : univ ⊆ ⋃ i, c i) : ∃ t : ℕ → Icc a b, t 0 = a ∧
+      Monotone t ∧ (∃ m, ∀ n ≥ m, t n = b) ∧ ∀ n, ∃ i, Icc (t n) (t (n + 1)) ⊆ c i := by
```

Comment:
Is this more or less convenient as:
```suggestion
    (hc₁ : ∀ i, IsOpen (c i)) (hc₂ : univ ⊆ ⋃ i, c i) :
    ∃ (n : ℕ) (t : Fin n.succ → Icc a b), t 0 = a ∧ t n = b ∧ Monotone t ∧ ∀ n, ∃ i, Icc (t n) (t (n + 1)) ⊆ c i := by
```

Reply:
Not sure what benefits that has; the current formulation is quite convenient for doing [induction](https://gist.github.com/alreadydone/c88265c99c73fa822d48e1abfd01282f#file-homotopylifting-lean-L207-L208) [over Nat](https://gist.github.com/alreadydone/c88265c99c73fa822d48e1abfd01282f#file-homotopylifting-lean-L281-L282).

Reply:
The main benefit is that the "junk" values after `t m` go away.

Reply:
And you can probably do one of the tuple inductions on `t`, but maybe it's not convenient

Reply:
You could argue that the last gap may not be the same as the others so there's "junk" there as well :) We could make it a finite strict monotone sequence by truncating, and if `α` is a field (or DivisibleBy Nat) we can get a genuinely *equally spaced* finite sequence with arbitrarily small gap, but I don't have uses for them. I consider this as a technical lemma so the important thing is that it applies easily (if stated in terms of Fin people would have to look into the less familiar Fin API), and people who need the stronger versions can get them by proving more about `addNsmul`.